File: arduinobot_firmware/arduinobot_firmware/simple_serial_transmitter.py
#!/usr/bin/python3
import rclpy
import serial
from rclpy.node import Node
from std_msgs.msg import String

class SimpleSerialTransmitter(Node):
    def __init__(self):
        super().__init__("simple_serial_transmitter")

        self.declare_parameter("port", "/dev/ttyUSB0")
        self.declare_parameter("baud_rate", 115200)
        self.port_ = self.get_parameter("port").value
        self.baud_rate_ = self.get_parameter("baud_rate").value

        self.sub_ = self.create_subscription(String, "serial_transmitter", self.msgCallback, 10)
        self.arduino_ = serial.Serial(port=self.port_, baudrate=self.baud_rate_, timeout=0.1)


    def msgCallback(self, msg):
        self.get_logger().info(f"New message received, publishing on port: {self.arduino_.name}")
        self.arduino_.write(msg.data.encode("utf-8"))



def main():
    rclpy.init()
    node = SimpleSerialTransmitter()
    rclpy.spin(node)
    node.destory_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: arduinobot_firmware/src/simple_serial_transmitter.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"
#include "libserial/SerialPort.h"

using std::placeholders::_1;

class SimpleSerialTransmitter : public rclcpp::Node
{
public:
    SimpleSerialTransmitter() : Node("simple_serial_transmitter")
    {
        declare_parameter<std::string>("port", "/dev/ttyUSB0");
        std::string port_ = get_parameter("port").as_string();
        
        subscription_ = this->create_subscription<std_msgs::msg::String>(
            "serial_transmitter", 10, std::bind(&SimpleSerialTransmitter::topic_callback, this, _1));
        
        arudino_.Open(port_);
        arudino_.SetBaudRate(LibSerial::BaudRate::BAUD_115200);
    }
    ~SimpleSerialTransmitter()
    {
        arudino_.Close();
    }

private:
  void topic_callback(const std_msgs::msg::String &msg)
  {
    RCLCPP_INFO(get_logger(), "New message received, publishing on serial port: %s", msg.data.c_str());
    arudino_.Write(msg.data.c_str());
  }
  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
  LibSerial::SerialPort arudino_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<SimpleSerialTransmitter>());
  rclcpp::shutdown();
  return 0;
}